### PR TITLE
Build: Upgrade Jest to version 28

### PR DESCRIPTION
**Pull request by Marklb**
Issue:

Angular tests have been excluded and they seem to be mostly fixed with Jest 28, so this is to try upgrading to Jest 28.

## What I did

This updates the monorepo to use jest 28.  Previously, we ran all tests in the `jsdom` environment, but the newer jest is more picky about it, and calls like `setImmediate` started to fail after the upgrade.  

So now we test some packages with a node environment and some with a jsdom environment. It does this by splitting the original jest config into two separate configs, one for browser environments, and one for node environments, and creating `jest.config.js` files in the root of each package.  It's a little tedious, but it gives us greater control over how jest runs for each package, not only the environment, but other config as well.

This also starts fixing Angular tests that have got behind while excluded, though we're not yet able to turn them back on just yet.

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by IanVS** [comment]
OK, I have tests _almost_ passing locally.  There's one problem I can't figure out, in storyshots, that I need some help on.  More on that later.

The main change I made here is to break out the jest config into two separate files, `jest.config.browser` and `jest.config.node`, both extending from a base config.  This is because we have some packages that are designed to run in node, and some in the browser, and previously we were running all our jest tests with the `jsdom` environment.  Well, it seems that jsdom has gotten more accurate now, so node.js APIs that we try to call in our node packages started throwing errors.

To accomplish this, I needed to start using a multi-project based approach, rather than multi-root.  Each project needs to have its own jest config, and these extend from either the browser config or the node config.  This also has two nice side-benefits: 

- We can configure jest differently for each project (for example, adding the `enzyme-to-json` serializer only for the storyshots addon, since that's the only place it's used.  
- We can give each project a name, which appears in the test when it's running.  (not a big deal, but kind of nice.)

Here's what the project names look like:
image

I'm not sure if I did a perfect job of breaking up node/browser packages, but the tests are passing so 🤷.  We can also specify a different environment on a per-file basis, as I did in `code/lib/api/src/tests/shortcut.test.js`.

OK so here's what I still need help with.  For some reason, a new snapshot is being created by the storyshots test, and it has the wrong path.  It seems to be due to `storybook.raw()` returning a `filename` that is relative, for the CSF story (but not the `storiesOf` story).  This is now being interpreted from the addons/storyshots directory rather than the project root.  Maybe @tmeasday has some idea of why this is happening?

And, I had to re-ignore the angular tests, because they were failing (and taking _forever_ to run).  @Marklb maybe you can take a look, by removing the exclusion in the `projects` glob in `jest.config.js`?

**Comment by tmeasday** [comment]
@ianvs

> Maybe @tmeasday has some idea of why this is happening?

I guess CSF files get a `fileName` that is a relative path (the same as the `entry.importPath`). This is assumed to be relative to the cwd where the webpack process runs from. Possibly that is interpreted by SS to be the Jest project root?

I'm not 100% sure what the `fileName` of a `storiesOf` file is, it's based on the `module.id`, but we don't actually use it directly (i.e import it) so it's not something I've looked at a lot. It could be absolute in Jest for sure.

**Comment by IanVS** [comment]
I'm not sure why the `bench` CI tasks are failing, they were passing in the commit prior to merging in `next`.  Other than that (possibly CI flake, though I retried several times), this PR is ready to review.

File: code/jest.config.base.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  'lib/core-common/src/utils/__tests__/template.test.ts',
+  'addons/storyshots/storyshots-core/src/frameworks/configure.test.ts',
+  'lib/core-common/src/utils/__tests__/interpret-files.test.ts',
+  'lib/builder-manager/src/utils/files.test.ts',
+  'lib/cli/src/helpers.test.ts',
+  'lib/core-server/src/utils/__tests__/server-statics.test.ts',
+  'lib/core-common/src/utils/__tests__/template.test.ts',
+  'addons/storyshots/storyshots-core/src/frameworks/configure.test.ts',
+  'lib/core-common/src/utils/__tests__/interpret-files.test.ts',
+  'lib/builder-manager/src/utils/files.test.ts',
```
Comment (ndelangen) [inline code/jest.config.base.js:19]: This array has been reduced by @valentinpalkovic in #19777

Comment (IanVS) [reply code/jest.config.base.js:19]: I guess it's a race to see which PR is merged first.  😅

File: code/jest.config.base.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    '/renderers/vue3/src/public-types.test.ts',
+    ...(process.platform === 'win32' ? skipOnWindows : []),
+  ],
+  collectCoverage: false,
+  collectCoverageFrom: [
+    'frameworks/*/src/**/*.{js,jsx,ts,tsx}',
+    'lib/*/src/**/*.{js,jsx,ts,tsx}',
+    'renderers/*/src/**/*.{js,jsx,ts,tsx}',
+    'addons/*/src/**/*.{js,jsx,ts,tsx}',
+  ],
```
Comment (ndelangen) [inline code/jest.config.base.js:74]:
```suggestion
    'ui/*/src/**/*.{js,jsx,ts,tsx}',
  ],
```

File: code/jest.config.base.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  coveragePathIgnorePatterns: [
+    '/node_modules/',
+    '/cli/test/',
+    '/dist/',
+    '/prebuilt/',
+    '/generators/',
+    '/template/',
+    '/dll/',
```
Comment (ndelangen) [inline code/jest.config.base.js:82]:
```suggestion
```

File: code/addons/a11y/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+const baseConfig = require('../../jest.config.browser');
+
+module.exports = {
+  ...baseConfig,
+  displayName: 'addons/a11y',
```
[the remark this answers is not in the document]
Comment (IanVS) [reply code/addons/a11y/jest.config.js:5]: We could I suppose.  Though honestly it's kind of nice to see the string there when you're editing a few of them, to remind which one you're looking at.  😅 

I wish I would have thought of this before copying/pasting/editing all these files, lol.  Right now I didn't add `lib/` to the core files, and left them as just their names.  I don't feel strongly how they should be named though, and could take the approach you suggest to make it easier to maintain in the future.

File: code/jest.init.base.ts [hunk 5]
```diff
@@ -58,41 +56,3 @@ const throwError = (message: any) => throwMessage('error: ', message);
...
-  }
-
-  addEventListener(event: string, cb: (data: any) => void) {
-    this.emitter.on(event, cb);
-  }
-
-  emit(event: string, data: any) {
-    this.emitter.emit(event, data);
-  }
-}
```
Comment (ndelangen) [inline code/jest.init.base.ts:96]: ❤️

File: code/jest.config.base.js (hunk 2 above)
Comment (IanVS) [reply code/jest.config.base.js:74]: I guess this means I should add a project for the `ui` folder as well, right?

File: code/jest.init.base.ts (hunk 5 above)
Comment (IanVS) [reply code/jest.init.base.ts:96]: It all just moved into the browser version, but it means that the node tests won't have these mocks.

File: code/lib/preview-web/src/PreviewWeb.test.ts [hunk 6]
```diff
@@ -2162,10 +2169,11 @@ describe('PreviewWeb', () => {
 
           // Wait three ticks without resolving the play function
           await waitForSetCurrentStory();
-          await waitForSetCurrentStory();
-          await waitForSetCurrentStory();
+          // We can't mock setTimeout for this test, due to waitForSetCurrentStory hack,
+          // So give some (real) time for the reload to be called
+          await new Promise((r) => setTimeout(r, 100));
```
Comment (ndelangen) [inline code/lib/preview-web/src/PreviewWeb.test.ts:2174]: This might turn out to be inconsistent?

**Review by ndelangen — COMMENTED** [review]
Are we upgrading to 28 for now, because of the storybook testrunner?

File: code/jest.init.base.ts (hunk 5 above)
Comment (ndelangen) [reply code/jest.init.base.ts:96]: right

File: code/jest.config.base.js (hunk 2 above)
Comment (ndelangen) [reply code/jest.config.base.js:74]: yes

File: code/jest.config.base.js (hunk 1 above)
Comment (ndelangen) [reply code/jest.config.base.js:19]: GO @valentinpalkovic GO!!!

File: code/addons/a11y/jest.config.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ndelangen) [reply code/addons/a11y/jest.config.js:5]: Sounds good!

**Comment by IanVS** [comment]
>Are we upgrading to 28 for now, because of the storybook testrunner?

No, this is just a big enough of a PR that I thought it might make sense to make this in two stages.  One PR to get to 28, then a subsequent PR to bump to 29 (which should be much easier to do, and will be a fast-follow).

File: code/lib/preview-web/src/PreviewWeb.test.ts (hunk 6 above)
Comment (IanVS) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2174]: Yeah I'm not sure.  Previously we were using a `setImmediate` to accomplish the same thing, so I _think_ we'll be safe, but I banged my head on this for half a day and couldn't come up with a better solution.

File: code/jest.config.base.js (hunk 1 above)
Comment (IanVS) [reply code/jest.config.base.js:19]: 🏎️

Comment (IanVS) [reply code/jest.config.base.js]: I went ahead and removed the same lines in this PR as well.

**Comment by IanVS** [comment]
This won't pass on windows until https://github.com/storybookjs/storybook/pull/19777 is merged.

File: code/lib/preview-web/src/PreviewWeb.test.ts (hunk 6 above)
Comment (tmeasday) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2193]: Hmm, do we want to disable this test for a second and I can take a look at it next week? It feels like it's getting pretty brittle and might need to be rethought.

Comment (tmeasday) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2193]: Or we could leave as is in this PR, and I can take a look then too.

Comment (IanVS) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2193]: It might be interesting to re-address once we get to jest 29, but I don't think anything will change.  

I think it would be great if you wanted to take a look at it when you've got some time, but it might be worth leaving the test running and skipping it only if we find out that it's flakey.  I don't feel strongly either way.

Comment (tmeasday) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2193]: Yes, let's leave it for now, I will make a note.

**Review by valentinpalkovic — APPROVED** [review]

**Comment by Marklb** [comment]
I am probably going to exclude https://github.com/storybookjs/storybook/blob/next/code/frameworks/angular/src/server/framework-preset-angular-cli.test.ts, because I think I can get the rest of Angular's tests working, but I am wondering if anyone has any ideas for how I can maybe solve the following issue blocking me on those tests. 

Here (https://github.com/storybookjs/storybook/blob/next/code/frameworks/angular/src/server/angular-cli-webpack-13.x.x.js#L27) we call `generateI18nBrowserWebpackConfigFromContext` and internally it is calling the same `loadEsmModule` function that we have defined here (https://github.com/storybookjs/storybook/blob/next/code/frameworks/angular/src/server/framework-preset-angular-ivy.ts#L21). When it attempts to import '@angular/core', which is published in ESM format only, a Segmentation fault is thrown. Everything I could find on the problem seems to be projects changing their code to avoid calling `loadEsmModule` during tests. We can't exactly change the code, because the function is in a node_modules package.

I started looking for a way to alter that line, using a custom transformer, but the errors looked like the dynamically imported module didn't go through the transformers. I may be able to alter it another way, so that it does transform the code correctly. Even if I can get that code to work by altering that line with a transformer, is it worth going for that approach or would it be better to mock that function and just make sure the rest of our code is doing what is expected?

**Comment by IanVS** [comment]
I'm so super confused by this.  Everything passes locally, `renderer/vue3` fails in the github actions CI jobs but passes in CircleCI, and `renderer/vue` fails in CircleCI, but passes in Github.  And everything was working just fine before I pulled in the latest `next` with the windows fixes.  Ahhhhhhh!

**Review by kasperpeulen — APPROVED** [review]
Amazing work!!

File: .circleci/config.yml [hunk 7]
```diff
@@ -176,22 +176,30 @@ commands:
             echo "Canceling workflow as previous step resulted in failure."
             echo "To execute all checks locally, please run yarn ci-tests"
             curl -X POST --header "Content-Type: application/json" "https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}/cancel?circle-token=${WORKFLOW_CANCELER}"
+  list-recent-merges:
```
Comment (kasperpeulen) [inline .circleci/config.yml:179]: @IanVS I guess this should not be part of this PR right?

Comment (IanVS) [reply .circleci/config.yml:179]: WTF, ugh.

File: MIGRATION.md [hunk 8]
```diff
@@ -865,47 +864,6 @@ import type { SvelteRenderer } from '@storybook/svelte';
 // etc.
 ```
 
-#### Renamed `DecoratorFn` to `Decorator`
```
[the remark this answers is not in the document]
Comment (IanVS) [reply MIGRATION.md]: Thanks, I wiped back to a previous commit, all the others were just failed experiments trying to get things to work, anyway.

**Comment by IanVS** [comment]
Phew, got the ✅ from CI finally.  I have a few approvals from folks, and the longer this sits the more I need to rebase it, so I'm going to merge, and start working on the jest 29 upgrade as well as a few other cleanups.  Thanks for the reviews, everyone, and thanks @Marklb for kicking this off.  Let's try to get angular tests passing and turned on in a separate PR.
